### PR TITLE
tests/common/lease: Wait for correct lease list response

### PR DESCRIPTION
**Pull request by endocrimes**
We don't consistently reach the same etcd server during the lifetime of
a test and in some cases, this means that this test will flake if an
etcd server was slow to update its state and the test hits the outdated
server.

Here we switch to using an `Eventually` case which will wait upto a
second for the expected result before failing - with a 10ms gap between
invocations.

```
[tests(dani/leasefix)] $ gotestsum -- ./common -tags integration -count 100 -timeout 15m -run TestLeaseGrantAndList
✓  common (2m26.71s)

DONE 1600 tests in 147.258s
```

### CODE REVIEW

**Comment by serathius** [comment]
Shouldn't `LeaseList` be a Linearizable read (preserve `read commited`)?

**Comment by endocrimes** [comment]
@serathius not as implemented in v3 afaict - https://github.com/etcd-io/etcd/blob/0d5c1dce49dd08cc833c7c5479fa67a0bd3c6e8c/server/etcdserver/v3_server.go#L364-L371

**Comment by ahrtr** [comment]
I think we should make the `LeaseList` as a Linearizable request, so this test case successfully detected the unexpected behavior. cc @ptabor

**Comment by endocrimes** [comment]
`LeaseLeases` is the only Lease API we have that doesn't get forwarded to the leader in some way, so that would make sense.

I assume it wasn't in the first place because it's a mostly informational api that's easy to overlook, but I can't find more discussion around it in the initial PR than this (https://github.com/etcd-io/etcd/pull/8358#issuecomment-322205538).

**Comment by ptabor** [comment]
Agree that intuitively `LeaseLeasesRequest` should be 'linearizable' 

I think it's pretty serious change to make it as part of stable v3.* line and violating formal backward compatibility guarantees to change the default. 

I would: 
  1. Add `bool linearizable = 1 [(versionpb.etcd_version_field)="3.6"];` to the LeaseLeases method (default false)
  2. Mark it as 'deprecated' 
  3. Add a new method `ListLeases`method with `serializable` field (default `false`): 
      https://github.com/etcd-io/etcd/blob/3d3c4373e3a4b029bd266afce307e69e3e399470/api/etcdserverpb/rpc.proto#L453

This way we would cleanup also the naming mistake (I assume) and have safe default going forward.

**Comment by endocrimes** [comment]
@ptabor Sounds good - I can open a PR for that soon. Do we wanna merge this to stabilize CI in the meantime (given that I assume a feature PR will take longer to land, and this is currently the flakiest etcd test)

**Comment by ahrtr** [comment]
I think it makes more sense to skip the test case for now instead of delivering a fix for an unexpected behavior.  Once the formal fix/enhancement is ready, we can enable it again.

File: tests/common/lease_test.go [hunk 1]
```diff
@@ -132,12 +132,21 @@ func TestLeaseGrantAndList(t *testing.T) {
...
+					// listing to eventually return true, either by the result propagaing
+					// or by hitting an up to date member.
+					leases := []clientv3.LeaseStatus{}
+					require.Eventually(t, func() bool {
+						resp, err := cc.LeaseList()
+						if err != nil {
+							return false
+						}
+						leases = resp.Leases
+						return len(resp.Leases) == nc.leaseCount
```
Comment (serathius) [inline tests/common/lease_test.go:145]: Checking if number of leases matches is not correct here as we cannot distinguish intermediate vs final state. I would recommend to instead check Revision. 

We can save revision from last `Grant` call and we we wait until we get a read with the revision. Then we can verify if the response has expected number of leases.

**Review by serathius — CHANGES_REQUESTED** [review]
Please use revision in eventually check

File: tests/common/lease_test.go (hunk 1 above)
Comment (endocrimes) [reply tests/common/lease_test.go:145]: sounds good 👍

Comment (endocrimes) [reply tests/common/lease_test.go]: apparently we do! - updated.

**Review by serathius — APPROVED** [review]

**Comment by ptabor** [comment]
> @ptabor Does adding `linearizable` field to `LeaseLeases` benefit us in any way when we will still introduce `ListLeases`? In both cases users will still take an action to get linealizable read (either set field or change method used). I think that adding the field will slow down support of new method. I think that we can skip adding `linearizable` field.

Seems good. 

> @endocrimes @ahrtr For now we are only deprecating `LeaseLeases`, we need to maintain the tests until we remove it. As so we still need to make this test stable.

**Pull request merged**